<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.0.38">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Yao-Chung Chen">
<meta name="dcterms.date" content="2022-08-15">

<title>Yao-Chung - TEKRABber</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../img/brain.ico" rel="icon">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-H0GJETSLQM"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-H0GJETSLQM', { 'anonymize_ip': true});
</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <a class="navbar-brand" href="../../index.html">
    <img src="../../img/brain.png" alt="">
    <span class="navbar-title">Yao-Chung</span>
  </a>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html">About</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../posts/posts.html">Posts</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../projects/projects.html">Projects</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://ferygood.github.io/resume/">Resume</a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ferygood"><i class="bi bi-github" role="img">
</i> 
 </a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/yao-chung-chen-3a0129173/"><i class="bi bi-linkedin" role="img">
</i> 
 </a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/YaoChungChen1"><i class="bi bi-twitter" role="img">
</i> 
 </a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="mailto:yaochung41@gmail.com"><i class="bi bi-envelope" role="img">
</i> 
 </a>
  </li>  
</ul>
              <div class="quarto-toggle-container">
                  <a href="" class="quarto-color-scheme-toggle nav-link" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
              </div>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#installation" id="toc-installation" class="nav-link" data-scroll-target="#installation">Installation</a></li>
  <li><a href="#comparing-between-two-species" id="toc-comparing-between-two-species" class="nav-link" data-scroll-target="#comparing-between-two-species">1. Comparing between two species</a></li>
  <li><a href="#ortholog-information-and-normalization" id="toc-ortholog-information-and-normalization" class="nav-link" data-scroll-target="#ortholog-information-and-normalization">1.1 Ortholog information and normalization</a></li>
  <li><a href="#inputs-for-differentially-expressed-de-and-correlation-analysis" id="toc-inputs-for-differentially-expressed-de-and-correlation-analysis" class="nav-link" data-scroll-target="#inputs-for-differentially-expressed-de-and-correlation-analysis">1.2 Inputs for differentially expressed (DE) and correlation analysis</a></li>
  <li><a href="#de-analysis" id="toc-de-analysis" class="nav-link" data-scroll-target="#de-analysis">1.3 DE analysis</a></li>
  <li><a href="#correlation-analysis" id="toc-correlation-analysis" class="nav-link" data-scroll-target="#correlation-analysis">1.4 Correlation analysis</a></li>
  <li><a href="#explore-results-apptekrabber" id="toc-explore-results-apptekrabber" class="nav-link" data-scroll-target="#explore-results-apptekrabber">1.5 Explore results <code>appTEKRABber()</code></a></li>
  <li><a href="#comparing-control-and-treatment-in-same-species" id="toc-comparing-control-and-treatment-in-same-species" class="nav-link" data-scroll-target="#comparing-control-and-treatment-in-same-species">2. Comparing control and treatment in same species</a></li>
  <li><a href="#de-analysis-1" id="toc-de-analysis-1" class="nav-link" data-scroll-target="#de-analysis-1">2.1 DE analysis</a></li>
  <li><a href="#correlation-analysis-1" id="toc-correlation-analysis-1" class="nav-link" data-scroll-target="#correlation-analysis-1">2.2 Correlation analysis</a></li>
  <li><a href="#explore-results-apptekrabber-1" id="toc-explore-results-apptekrabber-1" class="nav-link" data-scroll-target="#explore-results-apptekrabber-1">2.3 Explore results <code>appTEKRABber()</code></a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">TEKRABber</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Bioconductor</div>
    <div class="quarto-category">R</div>
    <div class="quarto-category">software</div>
    <div class="quarto-category">transposable elements</div>
    <div class="quarto-category">correlation</div>
  </div>
  </div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Yao-Chung Chen </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 15, 2022</p>
    </div>
  </div>
    
  </div>
  

</header>

<section id="introduction" class="level3">
<h3 class="anchored" data-anchor-id="introduction">Introduction</h3>
<p><a href="https://bioconductor.org/packages/release/bioc/html/TEKRABber.html">TEKRABber</a> is a Bioconductor package I published early this year. It is made to provide a user-friendly pipeline for <strong>comparing orthologs and transposable elements (TEs) between two species.</strong> It considers the orthology confidence between two species from BioMart to normalize expression counts and detect differentially expressed orthologs/TEs. Then it provides one to one correlation analysis for desired orthologs and TEs. There is also an app function to have a first insight on the result. Users can prepare orthologs/TEs RNA-seq expression data by their own preference to run <a href="https://bioconductor.org/packages/release/bioc/html/TEKRABber.html">TEKRABber</a> following the data structure mentioned in the following demo.</p>
<p>Although <a href="https://bioconductor.org/packages/release/bioc/html/TEKRABber.html">TEKRABber</a> is developed to compare between species, you can still use it for experiment which compare control and treatment in the same species. If you are more interest in this, you may directly jump to the second half of this article.</p>
</section>
<section id="installation" class="level3">
<h3 class="anchored" data-anchor-id="installation">Installation</h3>
<p>For a stable version, I recommend download <a href="https://bioconductor.org/packages/release/bioc/html/TEKRABber.html">TEKRABber</a> from Bioconductor archive.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"TEKRABber"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TEKRABber)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment) <span class="co"># load it if you are following this tutorial</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="comparing-between-two-species" class="level3">
<h3 class="anchored" data-anchor-id="comparing-between-two-species">1. Comparing between two species</h3>
<p>In the first example, we randomly generated/picked gene and transposable elements (TE) expression data from 10 humans and 10 chimpanzees <span class="citation" data-cites="khrameeva2020">(<a href="#ref-khrameeva2020" role="doc-biblioref">Khrameeva et al. 2020</a>)</span>. The values for the first column of gene and TE count table must be <strong>Ensembl gene ID</strong> and <strong>TE name</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load built-in data</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(speciesCounts)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>hmGene <span class="ot">&lt;-</span> speciesCounts<span class="sc">$</span>hmGene</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>hmTE <span class="ot">&lt;-</span> speciesCounts<span class="sc">$</span>hmTE</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>chimpGene <span class="ot">&lt;-</span> speciesCounts<span class="sc">$</span>chimpGene</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>chimpTE <span class="ot">&lt;-</span> speciesCounts<span class="sc">$</span>chimpTE</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># the first column must be Ensembl gene ID for gene, and TE name for TE</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(hmGene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           Geneid SRR8750453 SRR8750454 SRR8750455 SRR8750456 SRR8750457
1 ENSG00000000003        250        267        227        286        128
2 ENSG00000000005         13          2         15          9          5
3 ENSG00000000419        260        311        159        259        272
4 ENSG00000000457         86        131        100         94         80
5 ENSG00000000460         21         17         42         33         55
6 ENSG00000000938        162         75         95        252        195
  SRR8750458 SRR8750459 SRR8750460 SRR8750461 SRR8750462
1        394        268        102        370        244
2          0          1          8          0          2
3        408        371        126        211        374
4        158        119         46         77         81
5         29         50         11         18         20
6        137         93        108        197         69</code></pre>
</div>
</div>
</section>
<section id="ortholog-information-and-normalization" class="level3">
<h3 class="anchored" data-anchor-id="ortholog-information-and-normalization">1.1 Ortholog information and normalization</h3>
<p>In the first step, we use <code>orthologScale()</code> to get orthology information and calculate the scaling factor between two species. The species name needs to be the abbreviation of scientific species name used in Ensembl. (Note: (1)This step queries information using <a href="https://bioconductor.org/packages/3.15/biomaRt"><em>biomaRt</em></a> and it might need some time or try different mirrors due to the connections to Ensembl (2)It might take some time to calculate scaling factor based on your data size). For normalizing TEs, you need to provide a repeatmasker annotation table including four columns, (1) the name of TE (2) the class of TE (3) the average gene length of TE from your reference species (4) the average gene length from the species you want to compare. A way to download repeatmasker annotations is to query from <a href="http://genome.ucsc.edu/cgi-bin/hgTables?hgsid=1582443383_JGQc9hw61oGahqe20hrSWdfkjDWx&amp;clade=mammal&amp;org=Human&amp;db=hg38&amp;hgta_group=rep&amp;hgta_track=hg38&amp;hgta_table=0&amp;hgta_regionType=genome&amp;position=chr2%3A25%2C160%2C915-25%2C168%2C903&amp;hgta_outputType=primaryTable&amp;hgta_outFileName=hg38rmsk.csv">UCSC Genome Table Browser</a> and select the RepeatMasker track.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># You can use the code below to search for species name</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> biomaRt<span class="sc">::</span><span class="fu">useEnsembl</span>(<span class="at">biomart =</span> <span class="st">"genes"</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(biomaRt<span class="sc">::</span><span class="fu">listDatasets</span>(ensembl))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># In order to save time, we have save the data for this tutorial.</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(fetchDataHmChimp)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>fetchData <span class="ot">&lt;-</span> fetchDataHmChimp</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Query the data and calculate scaling factor using orthologScale():</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co">#' data(speciesCounts)</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="co">#' data(hg38_panTro6_rmsk)</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="co">#' hmGene &lt;- speciesCounts$hmGene</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="co">#' chimpGene &lt;- speciesCounts$chimpGene</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="co">#' hmTE &lt;- speciesCounts$hmTE</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="co">#' chimpTE &lt;- speciesCounts$chimpTE</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="co">#' </span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="co">#' ## For demonstration, here we only select 1000 rows to save time</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="co">#' set.seed(1234)</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="co">#' hmGeneSample &lt;- hmGene[sample(nrow(hmGene), 1000), ]</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="co">#' chimpGeneSample &lt;- chimpGene[sample(nrow(chimpGene), 1000), ]</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a><span class="co">#' </span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a><span class="co">#' fetchData &lt;- orthologScale(</span></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a><span class="co">#'     speciesRef = "hsapiens",</span></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a><span class="co">#'     speciesCompare = "ptroglodytes",</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a><span class="co">#'     geneCountRef = hmGeneSample,</span></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a><span class="co">#'     geneCountCompare = chimpGeneSample,</span></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a><span class="co">#'     teCountRef = hmTE,</span></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a><span class="co">#'     teCountCompare = chimpTE,</span></span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a><span class="co">#'     rmsk = hg38_panTro6_rmsk</span></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a><span class="co">#' )</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="inputs-for-differentially-expressed-de-and-correlation-analysis" class="level3">
<h3 class="anchored" data-anchor-id="inputs-for-differentially-expressed-de-and-correlation-analysis">1.2 Inputs for differentially expressed (DE) and correlation analysis</h3>
<p>We use <code>DECorrInputs()</code> to return inputs files for downstream analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>inputBundle <span class="ot">&lt;-</span> <span class="fu">DECorrInputs</span>(fetchData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="de-analysis" class="level3">
<h3 class="anchored" data-anchor-id="de-analysis">1.3 DE analysis</h3>
<p>In this step, we need to generate a metadata contain species name (i.e., human and chimpanzee). The row names need to be same as the DE input table and the column name must be <strong>species</strong> (see the example below). Then we use <code>DEgeneTE()</code> to perform DE analysis. When you are comparing samples between two species, the parameter <strong>expDesign</strong> should be <strong>TRUE</strong> (as default).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">species =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"human"</span>, <span class="fu">ncol</span>(hmGene) <span class="sc">-</span> <span class="dv">1</span>), </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rep</span>(<span class="st">"chimpanzee"</span>, <span class="fu">ncol</span>(chimpGene) <span class="sc">-</span> <span class="dv">1</span>))</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>meta<span class="sc">$</span>species <span class="ot">&lt;-</span> <span class="fu">factor</span>(meta<span class="sc">$</span>species, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"human"</span>, <span class="st">"chimpanzee"</span>))</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(meta) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(inputBundle<span class="sc">$</span>geneInputDESeq2)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>hmchimpDE <span class="ot">&lt;-</span> <span class="fu">DEgeneTE</span>(</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneTable =</span> inputBundle<span class="sc">$</span>geneInputDESeq2,</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">teTable =</span> inputBundle<span class="sc">$</span>teInputDESeq2,</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">metadata =</span> meta,</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">expDesign =</span> <span class="cn">TRUE</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="correlation-analysis" class="level3">
<h3 class="anchored" data-anchor-id="correlation-analysis">1.4 Correlation analysis</h3>
<p>Here we use <code>corrOrthologTE()</code> to perform correlation estimation comparing each ortholog and TE. This is the most time-consuming step if you have large data. For a quick demonstration, we use a relatively small data. You can specify the correlation method and adjusted <em>p-value</em> method. The default methods are Pearson’s correlation and FDR. <strong>Note:</strong> For more efficient and specific analysis, you can subset your data in this step to focus on only the orthologs and TEs that you are interested in.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load built-in data</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(speciesCorr)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>hmGeneCorrInput <span class="ot">&lt;-</span> <span class="fu">assay_tekcorrset</span>(speciesCorr, <span class="st">"gene"</span>, <span class="st">"human"</span>)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>hmTECorrInput <span class="ot">&lt;-</span> <span class="fu">assay_tekcorrset</span>(speciesCorr, <span class="st">"te"</span>, <span class="st">"human"</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>chimpGeneCorrInput <span class="ot">&lt;-</span> <span class="fu">assay_tekcorrset</span>(speciesCorr, <span class="st">"gene"</span>, <span class="st">"chimpanzee"</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>chimpTECorrInput <span class="ot">&lt;-</span> <span class="fu">assay_tekcorrset</span>(speciesCorr, <span class="st">"te"</span>, <span class="st">"chimpanzee"</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>hmCorrResult <span class="ot">&lt;-</span> <span class="fu">corrOrthologTE</span>(</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneInput =</span> hmGeneCorrInput,</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">teInput =</span> hmTECorrInput,</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">corrMethod =</span> <span class="st">"pearson"</span>,</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">padjMethod =</span> <span class="st">"fdr"</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>chimpCorrResult <span class="ot">&lt;-</span> <span class="fu">corrOrthologTE</span>(</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneInput =</span> chimpGeneCorrInput,</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">teInput =</span> chimpTECorrInput,</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">corrMethod =</span> <span class="st">"pearson"</span>,</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">padjMethod =</span> <span class="st">"fdr"</span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(hmCorrResult)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         geneName      teName      pvalue        coef      padj
1 ENSG00000143125        L1MD 0.271964872  0.38497828 0.9990235
2 ENSG00000143125        MSTA 0.335873091  0.34036703 0.9990235
3 ENSG00000143125  MLT1N2-int 0.966658172  0.01524552 0.9990235
4 ENSG00000143125       LTR57 0.067870603  0.59794954 0.9990235
5 ENSG00000143125 HERVK11-int 0.001028058  0.87118988 0.3210294
6 ENSG00000143125        LTR5 0.855235258 -0.06647109 0.9990235</code></pre>
</div>
</div>
</section>
<section id="explore-results-apptekrabber" class="level3">
<h3 class="anchored" data-anchor-id="explore-results-apptekrabber">1.5 Explore results <code>appTEKRABber()</code></h3>
<p><a href="https://bioconductor.org/packages/3.15/TEKRABber">TEKRABber</a> provides an app function for you to quickly view your result. First, you will need to assign the differentially expressed orthologs/TEs results, correlation results and metadata as global variables: <code>appDE</code>, <code>appRef</code>, <code>appCompare</code> and <code>appMeta</code>. See the following example.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#create global variables for app-use</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>appDE <span class="ot">&lt;-</span> hmchimpDE</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>appRef <span class="ot">&lt;-</span> hmCorrResult</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>appCompare <span class="ot">&lt;-</span> chimpCorrResult</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>appMeta <span class="ot">&lt;-</span> meta <span class="co"># this is the same one in DE analysis</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="fu">appTEKRABber</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/de.png" class="img-fluid"></p>
<p>The <strong>Expression tab</strong> show selected genes and TEs in boxplot. It also provide DE information and correlation statistics.</p>
<p><img src="images/corr.png" class="img-fluid"></p>
<p>In the <strong>Correlation tab</strong> page (above figure), you can select your data in scatter plots in three ways. <strong>(1)</strong> Specify the data point and it will turn red in the distribution of your results <strong>(2)</strong> Show the distribution of all the data from your correlation results based on their correlation coefficients and adjusted <em>p-value</em>. The blue vertical dashed line indicates the boundary of adjusted <em>p-value</em> is 0.05, and the orange one is for adjusted <em>p-value</em> 0.01 <strong>(3)</strong> you can click the data points which you are interested in, and it will be listed in the table. You can also drag a certain area to show data points in it.</p>
</section>
<section id="comparing-control-and-treatment-in-same-species" class="level3">
<h3 class="anchored" data-anchor-id="comparing-control-and-treatment-in-same-species">2. Comparing control and treatment in same species</h3>
<p>If you want to compare selected genes and TEs <strong>(1)</strong> from different tissue in same species or <strong>(2)</strong> control and drug treatment in same tissue in same species, please generate all the input files following the input format. Here we show an example data of prepared input files including expression counts from 10 control and 10 treatment samples. The format of input data: row names should be gene name or id, and column name is your sample id (please see details below).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load built-in data</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(ctInputDE)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>geneInputDE <span class="ot">&lt;-</span> ctInputDE<span class="sc">$</span>gene</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>teInputDE <span class="ot">&lt;-</span> ctInputDE<span class="sc">$</span>te</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="co"># you need to follow the input format as below</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(geneInputDE)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                control_1 control_2 control_3 control_4 control_5 treatment_6
ENSG00000180263      1470      2072      1864      2238      2246        2599
ENSG00000185985      1599      1045       946      1642      2199         665
ENSG00000144355       517       380      1211       812        48         388
ENSG00000234003         4         4        14        10         5           9
ENSG00000257342         1         1         1         2         3           3
ENSG00000223953       259       830       133       258       850         504
                treatment_7 treatment_8 treatment_9 treatment_10
ENSG00000180263        2679        2562        2532         2682
ENSG00000185985        1023        2477        1423         1731
ENSG00000144355         275         633          59          248
ENSG00000234003           4          18          13           22
ENSG00000257342           0           6           1            5
ENSG00000223953        1143        1500         498          864</code></pre>
</div>
</div>
</section>
<section id="de-analysis-1" class="level3">
<h3 class="anchored" data-anchor-id="de-analysis-1">2.1 DE analysis</h3>
<p>For DE analysis in the same species, you also use <code>DEgeneTE()</code> function, however, you need to set the parameter <strong>expDesign</strong> to <strong>FALSE</strong>. You also need to provide a metadata which this time the column name must be <strong>experiment</strong>. See demonstration below:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>metaExp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">experiment =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"control"</span>, <span class="dv">5</span>), <span class="fu">rep</span>(<span class="st">"treatment"</span>, <span class="dv">5</span>)))</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(metaExp) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(geneInputDE)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>metaExp<span class="sc">$</span>experiment <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    metaExp<span class="sc">$</span>experiment, </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"control"</span>, <span class="st">"treatment"</span>)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>resultDE <span class="ot">&lt;-</span> <span class="fu">DEgeneTE</span>(</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneTable =</span> geneInputDE,</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">teTable =</span> teInputDE,</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">metadata =</span> metaExp,</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">expDesign =</span> <span class="cn">FALSE</span></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="correlation-analysis-1" class="level3">
<h3 class="anchored" data-anchor-id="correlation-analysis-1">2.2 Correlation analysis</h3>
<p>For a quick demonstration to perform correlation of genes and TEs in control and treatment sample, we use relatively small input tables which only include 10 genes and 10 TEs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load built-in data</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(ctCorr)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>geneConCorrInput <span class="ot">&lt;-</span> <span class="fu">assay_tekcorrset</span>(ctCorr, <span class="st">"gene"</span>, <span class="st">"control"</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>teConCorrInput <span class="ot">&lt;-</span> <span class="fu">assay_tekcorrset</span>(ctCorr, <span class="st">"te"</span>, <span class="st">"control"</span>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>geneTreatCorrInput <span class="ot">&lt;-</span> <span class="fu">assay_tekcorrset</span>(ctCorr, <span class="st">"gene"</span>, <span class="st">"treatment"</span>)</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>teTreatCorrInput <span class="ot">&lt;-</span> <span class="fu">assay_tekcorrset</span>(ctCorr, <span class="st">"te"</span>, <span class="st">"treatment"</span>)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co"># you need to follow the input format as below</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(geneConCorrInput)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                control_1 control_2 control_3 control_4 control_5
ENSG00000180263      1470      2072      1864      2238      2246
ENSG00000185985      1599      1045       946      1642      2199
ENSG00000144355       517       380      1211       812        48
ENSG00000234003         4         4        14        10         5
ENSG00000257342         1         1         1         2         3
ENSG00000223953       259       830       133       258       850</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>controlCorr <span class="ot">&lt;-</span> <span class="fu">corrOrthologTE</span>(</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneInput =</span> geneConCorrInput,</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">teInput =</span> teConCorrInput,</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">corrMethod =</span> <span class="st">"pearson"</span>,</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">padjMethod =</span> <span class="st">"fdr"</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>treatmentCorr <span class="ot">&lt;-</span> <span class="fu">corrOrthologTE</span>(</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneInput =</span> geneTreatCorrInput,</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">teInput =</span> teTreatCorrInput,</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">corrMethod =</span> <span class="st">"pearson"</span>,</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">padjMethod =</span> <span class="st">"fdr"</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(treatmentCorr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         geneName teName    pvalue       coef      padj
1 ENSG00000180263    ALU 0.7350708 -0.2096204 0.8750842
2 ENSG00000180263  AluJb 0.6463988 -0.2814802 0.8729353
3 ENSG00000180263  AluJo 0.5879945 -0.3296607 0.8399922
4 ENSG00000180263  AluJr 0.6080013 -0.3130670 0.8563399
5 ENSG00000180263 AluJr4 0.7158070 -0.2251209 0.8729353
6 ENSG00000180263  AluSc 0.7130150 -0.2273721 0.8729353</code></pre>
</div>
</div>
</section>
<section id="explore-results-apptekrabber-1" class="level3">
<h3 class="anchored" data-anchor-id="explore-results-apptekrabber-1">2.3 Explore results <code>appTEKRABber()</code></h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>appDE <span class="ot">&lt;-</span> resultDE</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>appRef <span class="ot">&lt;-</span> controlCorr</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>appCompare <span class="ot">&lt;-</span> treatmentCorr</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>appMeta <span class="ot">&lt;-</span> metaExp</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">appTEKRABber</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/corr2.png" class="img-fluid"></p>
</section>
<section id="references" class="level3">
<h3 class="anchored" data-anchor-id="references">References</h3>
<ol type="1">
<li><p>TEKRABber on Bioconductor: <a href="https://bioconductor.org/packages/release/bioc/html/TEKRABber.html">Link</a></p></li>
<li><p>Source code on Github: <a href="https://github.com/ferygood/TEKRABber">Link</a></p></li>
</ol>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-khrameeva2020" class="csl-entry" role="doc-biblioentry">
Khrameeva, Ekaterina, Ilia Kurochkin, Dingding Han, Patricia Guijarro, Sabina Kanton, Malgorzata Santel, Zhengzong Qian, et al. 2020. <span>“Single-cell-resolution transcriptome map of human, chimpanzee, bonobo, and macaque brains.”</span> <em>Genome Research</em> 30 (5): 776–89. <a href="https://doi.org/10.1101/gr.256958.119">https://doi.org/10.1101/gr.256958.119</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } 
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
<script src="https://giscus.app/client.js" data-repo="ferygood/ferygood.github.io" data-repo-id="R_kgDOGgjrTg" data-category="General" data-category-id="DIC_kwDOGgjrTs4CQ5S1" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
      <div class="nav-footer-center">Yao-Chung Chen © All Rights Reserved. 2023</div>
  </div>
</footer>



</body></html>